docs(css-easing-function): Rewrite the "Values" section (#29585)
mdn/content: files/en-us/web/css/easing-function/index.md

/* cubic-bezier function and keywords */
/* cubic-bezier(<x1>, <y1>, <x2>, <y2>) */

> cubic-bezier(0.42, 0, 1, 1)

Author: Thanks, I've updated the syntax example to `cubic-bezier(0.25, 0.1, 0.25, 1)`.

> might as well add ease-in-out to the list on line 28, that way they're all listed.

`ease-in-out` exists in the list on line 28, unless I am missing something

### Values

- `linear`

>   - : Specifies a constant rate of interpolation, ensuring there is no change in the rate of progress of the animation throughout the duration. This also means that there is no acceleration or deceleration at any point during the animation's duration. It's important to not confuse this keyword value with the [linear() easing function](#linear_easing_function), although they are related: the `linear` keyword produces a `linear()` easing function with two points, `linear(0, 1)`.

Author: Updated as:

> Specifies a constant rate of interpolation, with no change in the rate of progress throughout the duration (that is, no acceleration or deceleration). This keyword value is equivalent to both [`cubic-bezier(0, 0, 1, 1)`](#cubic_bézier_easing_function) and [`linear(0, 1)`](#linear_easing_function).

>     > **Note:** The `linear` keyword always serializes as `linear`, whereas its function equivalent `linear(0, 1)` serializes as `linear(0 0%, 1 100%)`.

Author: I was basing these updates on the `linear` keyword is explained in the spec: https://drafts.csswg.org/css-easing/#valdef-easing-function-linear.

To keep this keyword description simple as suggested (I agree), I've moved the Note to the [Linear easing function](https://pr29585.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/easing-function#linear_easing_function) section. Also simplified "serializes" to "interpreted as":

> **Note:** The [`linear`](#linear) keyword is equivalent to the easing function `linear(0, 1)`. While the `linear` keyword is always interpreted as-is, the function `linear(0, 1)` gets interpreted as `linear(0 0%, 1 100%)`.

![Graph of "Input progress" to "Output progress" showing a line extending from the origin to (1, 1).](linear.svg)

- `<linear-easing-function>`

  - : Specifies a `linear()` function using multiple points, called linear stops, allowing you to control the movement of an animation at specific points along the path.

>     The `linear()` function specifies a list of points (`<linear-stop-list>`), separated by commas. Each point or linear stop in the list (`<linear-stop>`) is specified as a {{cssxref("&lt;number&gt;")}} and optional one or two {{cssxref("&lt;percentage&gt;")}} values (`<linear-stop-length>`).

Author: Thanks for helping with simplifying Updated this to:

> Specifies a `linear()` function with one or more comma-separated _linear stops_, each containing up to two optional _stop lengths_, to control the progress of an animation or transition.

>  The `linear()` function specifies a `<linear-stop-list>`, a comma-separated list of points along the animation or transition progress. Each point or `<linear-stop>` in the list is specified as a {{cssxref("&lt;number&gt;")}} between `0` and `1`. By default, each stop in the `<linear-stop-list>` is equidistant. To have more control on the progress of the animation or transition, each point can include up to two optional {{cssxref("&lt;percentage&gt;")}} `<linear-stop-length>` values.

    - `<number>`: Represents a point in time. The value should be between `0` and `1`.

    - `<percentage>`: Represents the progress of the animation at the specified point in time. If two percentage values are specified, they represent a range for that specific point – the first percentage indicates the starting point and the second percentage indicates the ending point of the progress for that segment in the animation. If no `<percentage>` value is specified, the points are spread evenly along the timeline.

- `<cubic-bezier-easing-function>`

>   - : Specifies a `cubic-bezier()` function using two control points, which help shape the curve of the animation's progress. This parameter accepts one of the following four keyword values that map to predefined `cubic-bezier()` functions or a custom `cubic-bezier()` function to specify the control points:

Author: Thanks for the hints! I've rephrased this as:

> Specifies a `cubic-bezier()` function using two control points, each with two values, which help shape the curve of the animation or transition progress. The control points can be specified using a custom `cubic-bezier()` function, or one of the four predefined keyword values for the `cubic-bezier()` function can be used. The predefined keywords include:

    `ease-out`: This keyword represents the easing function `cubic-bezier(0, 0, 0.58, 1)`. It indicates that the interpolation starts abruptly and then progressively slows down towards the end.

    `ease-in-out`: This keyword represents the easing function `cubic-bezier(0.42, 0, 0.58, 1)`. It indicates that the interpolation starts slowly, speeds up, and then slows down towards the end. At the beginning, it behaves like the `ease-in` keyword; at the end, it is like the `ease-out` keyword.

    ![Graphs of "Input progress" to "Output progress", of which "ease" shows a curved line quickly rising from the origin to (1, 1); "ease-in" shows a shallow curved line from the origin that straightens out as it approaches (1, 1); "ease-out" shows a straight diagonal line that slightly curves as it gets close to (1, 1); and "ease-in-out" shows a symmetrical, "S"-shaped line curving from the origin to (1, 1).](ease.svg)

    `cubic-bezier()`: This function accepts four {{cssxref("&lt;number&gt;")}} values.

>     - `<number>`: Specifies the abscissas and ordinates of the [P1 and P2 points](#cubic_bézier_easing_function) of the curve (as `<x1>`, `<y1>`, `<x2>`, and `<y2>`). `<x1>` and `<x2>` must be in the range `[0, 1]`, otherwise the value is invalid.

Author: 🙂 How does this look:

> - {{cssxref("&lt;number&gt;")}}: Specifies the location of [P1 and P2 points](#cubic_bézier_easing_function) on the curve. `<x1>` and `<y1>` are the coordinates for point P1, and `<x2>` and `<y2>` are the coordinates for point P2. `<x1>` and `<x2>` values must be between `0` and `1`, otherwise the function will not work as expected.

  - : Specifies a `steps()` function that divides the animation into a set number of intervals or "steps", causing the animation to jump from one step to the next rather than transitioning smoothly. This parameter accepts one of the following two keyword values that map to predefined `steps()` functions or a custom `steps()` function:

    `step-start`: This keyword represents the easing function `steps(1, jump-start)` or `steps(1, start)`. It indicates that the interpolation jumps immediately to its final state, where it stays until the end.

    `step-end`: This keyword represents the easing function `steps(1, jump-end)` or `steps(1, end)`. It indicates that the interpolation stays in its initial state until the end, at which point it jumps directly to its final state.

    ![Graphs of "Input progress" to "Output progress", of which "step-start" shows a hollow origin and a horizontal line extending from (0, 1) to (1, 1); and "step-end" shows a horizontal line extending from the origin to (1, 0) (hollow) and a point at (1, 1).](step.svg)

    `steps()`: This function accepts a strictly positive {{cssxref("&lt;integer&gt;")}} and a second optional `<step-position>`.

>     - `<integer>`: Represents the number of equidistant intervals or 'steps' in the animation. It must be a positive integer greater than `0` unless the second parameter is `jump-none`, in which case, it must be a positive integer greater than `1`.

Reviewer: ```suggestion
    - `<integer>`: Represents the number of equidistant intervals or 'steps'. It must be a positive integer greater than `0` unless the second parameter is `jump-none`, in which case, it must be a positive integer greater than `1`.
```

>     - `<step-position>`: Indicates when the animation jumps from one step to the next. The possible keyword values include:

Reviewer: ```suggestion
    - `<step-position>`: Indicates specifies the timing of the first jump, last jump, or both. The possible keyword values include:
```
What i wrote might not be perfect, but the step position is about the first and last jumps, not each steps timing, which is what the original seems to indicate.

Author: Thanks for pointing out. Updating it to:

> Specifies the timing of the jump either at the start, the end, or both.

>       - `jump-start` denotes that the first step or jump happens when the interpolation begins.

Reviewer: ```suggestion
      - `jump-start` denotes that the first step or jump happens when the interpolation begins, essentially jumping past the `0` point.
```

Author: Thanks for the hint. Updating this to:

> - `jump-start` denotes that the first step occurs right at the start, essentially beginning at the `0` point.

>       - `jump-end` denotes that the last step or jump happens when the interpolation ends. This is the default value when no `<step-position>` is specified.

Reviewer: ```suggestion
      - `jump-end` denotes that the last step or jump happens when the interpolation ends, staying on `0%` but no duration of time at `100%`. This is the default value when no `<step-position>` is specified.
```

Author: To keep the structure similar to that of `jump-start`, how about rewording `jump-end` to:

> - `jump-end` denotes that the last step occurs right at the end, essentially jumping at the `1` point at the end of the duration. This is the default value when no `<step-position>` is specified.

      - `jump-none` denotes that no jump happens at the beginning or the end, effectively removing a step during the interpolation iteration. Instead, the animation holds steady at both the 0% mark and the 100% mark. The duration of these holds is determined by dividing the total animation duration by the number of steps (1/n).

>       - `jump-both` denotes that jumps occur at both the 0% and 100% marks, effectively adding a step during the interpolation iteration.

Reviewer: ```suggestion
      - `jump-both` denotes that jumps occur at both ends, spending to time at the 0% and 100% marks, effectively adding a step during the interpolation iteration.
```

Author: Updating this one to:

>  - `jump-both` denotes that jumps occur at both the beginning and the end, essentially occurring at both the `0` and `1` points. Effectively, a step is added at both the ends. No time is spent at the `0%` and `100%` marks.

    - `<number>`: Represents a point in time along the duration of the animation or transition. The value should be between `0` and `1`, where `0` represents the start of the iteration and `1` represents the end.

    - `<percentage>`: Indicates the position of a linear stop along the duration. It can take up to two values. If one value is specified, it defines the start of the associated linear stop. If two percentage values are specified, they define the length of the stop: the first percentage indicates the starting point and the second percentage indicates the ending point for that segment in the animation or transition. If no `<percentage>` value is specified, which is the default, the stops are spread evenly along the timeline.

- `<cubic-bezier-easing-function>`

>   - : Specifies a `cubic-bezier()` function using two control points, each with two values, which help shape the curve of the animation or transition progress. The control points can be specified using a custom `cubic-bezier()` function, or one of the four predefined keyword values for the `cubic-bezier()` function can be used. The predefined keywords include:

Reviewer: ```suggestion
  - : Specifies a Bézier curve that shapes the progression of the animation or transition.  In CSS, [Bézier curves] are mathematically described curves defined by four control points. The cubic bezier easing function is defined either via either one of the keyword values mapped to common bezier curve parameter values or, to create a custom curve, via a four-parameter `cubic-bezier()` function call. The four predefined keywords include:
```

Author: Thanks! The suggestion adds more clarity. I've updated the text as:

> - : Specifies a [Bézier curve](/en-US/docs/Glossary/Bezier_curve) to shape the progress of an animation or a transition. In CSS, Bézier curves are defined by four control points that mathematically describe the curve: a starting point, an ending point, and two control points. The cubic Bézier easing function can be defined in one of these two ways: by creating a custom curve with a four-parameter `cubic-bezier()` function call or by using one of the predefined keyword values, which map to the commonly used Bézier curve parameters. The predefined keyword values include:

Also added link to glossary here. The term is linked to glossary in "Description" > "Cubic Bézier easing function" section as well.
